libc/src/ringbuffer.c

```diff
@@ -0,0 +1,96 @@
+#include <ringbuffer.h>
+#include <stdlib.h>
+
+/* Initialize a ringbuffer for use takes a pointer to a declared ring buffer
+ * and initializes it with size size it will be allocated in size * isz
+ * returns: 0 on success -1 on failure, if this fails consider the ringbuffer
+ * pointed at by ref invalid and unsafe to use errno is likely to have a cause
+ * for the failure
+ */
+int ringbuffer_init(ringbuffer_t* ref, size_t size) {
+    ref->size = size;
+    ref->unread_data = 0;
+
+    ref->data = (uint8_t*) malloc(size);
+
+    if (ref->data == NULL) {
+        return -1;
+    }
+
+    ref->r_pos = 0;
+    ref->w_pos = 0;
+
+    return 0;
+}
+
+/* Allocate and initialize a ringbuffer returns NULL on failure to allocate */
+ringbuffer_t* ringbuffer_new(size_t size) {
+    ringbuffer_t* ref = malloc(sizeof(ringbuffer_t));
+    if (ringbuffer_init(ref, size) == -1) {
+        free(ref);
+        ref = NULL;
+    }
+
+    return ref;
+}
+
+size_t ringbuffer_used(ringbuffer_t* ref) {
+    return ref->unread_data;
+}
+
+/* How much space is available before the writing more data would overwrite data
+ * that has not been read yet
+ */
+size_t ringbuffer_available(ringbuffer_t* ref) {
+    return ref->size - ref->unread_data;
+}
+
+void ringbuffer_free(ringbuffer_t* ref) {
+    free(ref->data);
+    free(ref);
+}
+
+/* Write n bytes to the ringbuffer pointed at by ref returns 0 if the data did
+ * fit, 1 if old data was overwritten or -1 if there is insufficient space before
+ * the read pointer to place the data
+ */
+int ringbuffer_write(ringbuffer_t* ref, size_t n, uint8_t* data) {
+    uint8_t* pos = ref->data + ref->w_pos;
```

Comment:
Unused variable. The code should compile with no warnings at all.

Reply:
found the strict target on make I will comb the code with it those ended leftovers after I changed how the arithmetic is done